[NU-1904] Add possiblity to configure deployment parameters for Eager…

lukasz-bigorajski opened the pull request:
## Describe your changes
![image](https://github.com/user-attachments/assets/fee8f356-8528-439d-9cac-29e90ebecc32)

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

gskrobisz commented on extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessAction.scala:
```diff
@@ -109,8 +109,8 @@ object ScenarioActionName {
 
 /**
- * Used to define Source parameters for each action
+ * Used to define Flink deployment parameters for each action
```
This is not clear for me. I guess action deploy in Flink environment is one of the possible use cases for this configuration.

lukasz-bigorajski replied:
Removed 'deployment' - done

gskrobisz commented on engine/flink/management/dev-model/src/main/scala/pl/touk/nussknacker/engine/management/sample/LoggingService.scala:
```diff
@@ -31,19 +36,39 @@ object LoggingService extends EagerService {
         Future.successful(())
       }
 
+      private def isNotDisabled(componentUseContext: ComponentUseContext) =
```
Sounds like double negative. Do you mean "isEnabled"?

lukasz-bigorajski replied:
It was for a purpose, but I refactored it a little

gskrobisz commented on engine/flink/management/dev-model/src/main/scala/pl/touk/nussknacker/engine/management/sample/LoggingService.scala:
```diff
@@ -31,19 +36,39 @@ object LoggingService extends EagerService {
+
+      override def actionParametersDefinition: Map[ScenarioActionName, Map[ParameterName, ParameterConfig]] = Map(
+        ScenarioActionName.Deploy -> Map(
+          ParameterName("loggingComponentsDisabled") -> ParameterConfig(
```
How about inverting logic? When I insert logging component to my scenario diagram I expect that logging is being done. Then, when I invoke "deploy" action, I can decide to supress logging.  

EDIT: I see that default value is "false" (and it IS logging by default), but I am sure that does not help to comprehend what is going on here.

lukasz-bigorajski replied:
Done

gskrobisz commented on components/openapi/src/it/scala/pl/touk/nussknacker/openapi/functional/OpenAPIServiceSpec.scala:
```diff
@@ -31,9 +31,9 @@ class OpenAPIServiceSpec
-  implicit val componentUseCase: ComponentUseCase = ComponentUseCase.EngineRuntime
-  implicit val metaData: MetaData                 = MetaData("testProc", StreamMetaData())
-  implicit val context: Context                   = Context("testContextId", Map.empty)
+  implicit val componentUseCase: ComponentUseContext = ComponentUseContext.EngineRuntime(None)
```
Rename componentUseCase to componentUseContext

lukasz-bigorajski replied:
Done

Commit:
[NU-1904] Add possiblity to configure deployment parameters for EagerService

# Conflicts:
#	benchmarks/src/test/scala/pl/touk/nussknacker/engine/benchmarks/interpreter/InterpreterSetup.scala
#	components/openapi/src/it/scala/pl/touk/nussknacker/openapi/functional/OpenAPIServiceSpec.scala
#	components/openapi/src/test/scala/pl/touk/nussknacker/openapi/BaseOpenAPITest.scala
#	components/sql/src/main/scala/pl/touk/nussknacker/sql/service/DatabaseEnricherInvoker.scala
#	designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/ActionInfoService.scala
#	designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ActionInfoResourcesSpec.scala
#	docs/MigrationGuide.md
#	engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkCustomNodeContext.scala
#	engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/FlinkProcessCompilerData.scala
#	engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/FlinkProcessCompilerDataFactory.scala
#	engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/StubbedFlinkProcessCompilerDataFactory.scala
#	engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/TestFlinkProcessCompilerDataFactory.scala
#	engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/VerificationFlinkProcessCompilerDataFactory.scala
#	engine/flink/management/dev-model/src/main/scala/pl/touk/nussknacker/engine/management/sample/service/DynamicMultipleParamsService.scala
#	engine/lite/request-response/runtime/src/test/scala/pl/touk/nussknacker/engine/requestresponse/RequestResponseInterpreterSpec.scala
#	engine/lite/runtime/src/main/scala/pl/touk/nussknacker/engine/lite/ScenarioInterpreterFactory.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/ComponentExecutorFactory.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompilerData.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/NodeCompiler.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/NodeDataValidator.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compiledgraph/service.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/ModelDataActionInfoProvider.scala
#	scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/component/methodbased/CustomStreamTransformerExtractor.scala
#	scenario-compiler/src/test/scala/pl/touk/nussknacker/engine/InterpreterSpec.scala
#	scenario-compiler/src/test/scala/pl/touk/nussknacker/engine/compile/ProcessValidatorSpec.scala
#	scenario-compiler/src/test/scala/pl/touk/nussknacker/engine/testcomponents/SpelTemplatePartsService.scala
#	utils/components-utils/src/main/scala/pl/touk/nussknacker/engine/util/service/EagerServiceWithStaticParameters.scala
#	utils/lite-components-testkit/src/test/scala/pl/touk/nussknacker/engine/lite/util/test/LiteTestScenarioRunnerSpec.scala
#	utils/lite-components-testkit/src/test/scala/pl/touk/nussknacker/engine/lite/util/test/TestService.scala

Commit:
scala fmt

arkadius commented on components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ComponentUseContext.scala:
```diff
@@ -0,0 +1,32 @@
+
+  def deploymentData(): Option[NodeDeploymentData] = this match {
+    case EngineRuntime(nodeData) => nodeData
+    case _                       => None
```
Let's use explicit cases matching. Thanks to that, we avoid some mistakes during changes around these contexts. Also, why not a normal polymorphism? I mean `def deploymentData`

lukasz-bigorajski replied:
Done

arkadius replied:
Scala recommend to use braceless methods (`def deploymentData: Option[NodeDeploymentData`) for methods that have no side effects

lukasz-bigorajski replied:
Methods without brackets are confusing for me because when invoke it I don't whether it's a variable or a method. This is why I chose method with brackets. So I'm really wondering whether to remove it

arkadius replied:
I agree that it is a tricky conceptual decision made by creators of scala, but IMO being consistent with standards is very important value. We could talk about changing all braceless methods to methods with brackets in all scala projects in our ecosystem, but I think that it is not worth it ATM

arkadius commented on components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ComponentUseContext.scala:
```diff
@@ -0,0 +1,32 @@
+object ComponentUseContext {
+  case class EngineRuntime(nodeData: Option[NodeDeploymentData]) extends ComponentUseContext
+  case object TestRuntime                                        extends ComponentUseContext
+  case object Validation                                         extends ComponentUseContext
```
Can you remove contexts:
- `ServiceQuery` - It was for purpose of external project, I checked their repo and found that they can replace it with `EngineRuntime(None)` without any damage
- `Validation` - Our validation mechanism compiles components, and after that regrets result of compilation so, these components never will be used - we can pass `EngineRuntime(None)` there with some additional note that it is not important
- `TestDataGeneration` - During test data generation things are done very similar to validation. We compile components only to get test data generator, so component is not used for anything else and it doesn't need to know the context.
?

Also, I would rename two other contexts:
- `EngineRuntime` - scenario testing mechanism also run component on some engine so it is confusing, I would rename it to `LiveDeployment`/`RealDeployment` 
- `TestRuntime` - `ScenarioTesting`. Meantime we clarified this name for naming this thing. We have `scenarioTesting` in endpoints and in other places in the code

lukasz-bigorajski replied:
Done - I renamed 'EngineRuntime' to 'LiveRuntime', what do you think?

arkadius replied:
LGTM, nice

arkadius commented on scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala:
```diff
@@ -27,17 +31,57 @@ abstract class CommonModelDataInfoProvider(modelData: ModelData) {
+      compileSourcesAndService(scenario.collectAllNodes.flatMap(asFragmentInputDefinition))
   }
 
+  private def compileSourcesAndService(nodes: List[NodeData])(implicit jobData: JobData): Map[NodeComponentInfo, Any] =
```
I think, that we should compile all non-built-in nodes. Consider situations:
- custom node - it is more general case as service. it can be also configurable
- sinks - I could imagine some situation when you want to specify behaviour of sink during deployment
We could introduce a sealed trait for that types of components, I mean `NonBuiltInComponent`. `CustomComponent` would be better but it is already reserved for `CustomNode`.

lukasz-bigorajski replied:
I fixed others, but I have to find out how to compile all nodes at once

arkadius replied:
You don't need to compile all, built-in components can't be ommitted

arkadius reviewed (commented):
Looks good, some minor comments added

Commit:
review fixes

arkadius commented on engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/MetricsProviderForFlink.scala:
```diff
@@ -1,18 +1,18 @@
-    componentUseCase match {
-      case ComponentUseCase.TestRuntime => NoOpMetricsProviderForScenario
-      case _                            => new FlinkMetricsProviderForScenario(runtimeContext)
+    componentUseContextProvider match {
```
Hmm I see that we use this class not only to provide `ComponentUseContext` but also for some pattern matching for other mechanism. I think that we should revert the name to sth like `RuntimeMode` or sth like that.

lukasz-bigorajski replied:
I like the name

Commit:
more review fixes

Commit:
subgraph compilation

Commit:
review fixes

Commit:
review fixes 2

Commit:
test fix

arkadius commented on scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala:
```diff
@@ -26,18 +31,65 @@ abstract class CommonModelDataInfoProvider(modelData: ModelData) {
 
-  protected def collectAllSources(scenario: CanonicalProcess): List[SourceNodeData] = {
-    scenario.collectAllNodes.flatMap(asSource) ++ scenario.collectAllNodes.flatMap(asFragmentInputDefinition)
+  def compileScenario(scenario: CanonicalProcess)(implicit jobData: JobData): Map[NodeComponentInfo, Any] = {
```
I think that it is important to emphasize somehow (maybe in method name?) that the purpose of this compilation is collection of all custom nodes (not filter or variable). Looking at current name it looks like we compile scenario to some executable unit. Maybe let's rename it to `compileAllCustomNodes`? Also this fallback to `Map.empty` is tricky. Maybe let's move it to place of invocation for clarity?

lukasz-bigorajski replied:
Done

arkadius commented on scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala:
```diff
@@ -26,18 +31,65 @@ abstract class CommonModelDataInfoProvider(modelData: ModelData) {
+      }
+    }
+
+  private def compilePart(part: ProcessPart)(implicit jobData: JobData): Map[NodeComponentInfo, Any] =
```
The same as above, we need to emphasize that this compilation is for the purpose of collection of all custom nodes. Maybe the same name `compileAllCustomNodes`? Or `compileAllPartCustomNodes`

lukasz-bigorajski replied:
Done

arkadius commented on scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/compiledgraph/CompiledNodesCollector.scala:
```diff
@@ -0,0 +1,38 @@
+
+object CompiledNodesCollector {
+
+  def collectNodes(node: Node): List[Node] = {
```
maybe `collectAllNodes` will be a little bit less confusing?

lukasz-bigorajski replied:
Done

arkadius reviewed (commented):
Nice, minor comments added, but in general looks good

Commit:
review fixes

arkadius reviewed (approved):
Good job

Commit:
update swagger docs
